lua/luasnip/extras/fmt.lua

```diff
@@ -0,0 +1,159 @@
+local ls = require('luasnip')
+
+-- Format a string using a syntax similar to Python ''.format() method.
+--
+-- Placeholders are enclosed in `{}` and can be:
+-- * with numbers, e.g. `{1}` - takes `args` elemengs by index, e.g. `args[1]`
+-- * with names, e.g. `{myname}` - takes `args` elements by key, e.g. `args['myname']`
+-- * empty, e.g. `{}`, like with numbers, but takes the next index, if no numbered
+--   placeholders have been used, then it starts indexing from 1, else the value
+--   of last numbered placeholder + 1 is used
+-- To input regular curly braces use `{{}}` or `{{text}}`. The text between braces
+-- is not matched by [0-9a-zA-Z_-]* then single braces can be used, no need to escape.
+--
+-- Arguments:
+--   fmt: string
+--   args: table
+-- Returns: a list of strings and elements of `args` inserted into placeholders.
+local function replace_placeholders(fmt, args)
+    local pattern = '{[0-9a-zA-Z_-]*}'
+    local elements = {}
+
+    local add_text = function(text)
+        if #text > 0 then
+            table.insert(elements, text)
+        end
+    end
+    local add_arg = function(key)
+        assert(args[key], string.format('Missing key `%s` in format arguments.', key))
+        table.insert(elements, args[key])
+    end
+
+    local last_index = 0
+    local start = 0
+    local end_ = 0
+    local prev_end = 0
+
+    while true do
+        start, end_ = fmt:find(pattern, end_+1)
+        -- if not found, add the remaining part of string and finish
+        if start == nil then
+            add_text(fmt:sub(prev_end + 1))
+            break
+        end
+        -- ignore non-placeholders, i.e. double curly braces replacing them with
+        -- single braces, e.g. `{{2}}` -> `{2}`
+        local before = fmt:sub(start - 1, start)
+        local after = fmt:sub(end_, end_ + 1)
+        local ignored = before == '{{' and after == '}}'
+        if ignored then
+            -- add the previous part of the string
+            add_text(fmt:sub(prev_end + 1, start - 1))
+            -- add what is inside {{}}
+            add_text(fmt:sub(start + 1, end_ - 1))
+            -- move previous end
+            prev_end = end_
+        else
+            -- add the previous part of the string
+            add_text(fmt:sub(prev_end + 1, start - 1))
```

Comment:
The two `add_text()` could be moved out of the `if`, right?

Reply:
Yes, I'll change that.